### DISCUSSION TITLE
"Cannot access member" when overloading/rewriting an inherited class method of an included module?

### DISCUSSION THREAD

**Discussion by HunterAP23** [microsoft/pyright#1991] [Q&A] [answered]
Mostly unsure what's causing this, but I figured I could put it up to discussion.

I have a class called `pool_test` that uses `ProcessPoolExecutor` from the `concurrent.futures` module as a parent. I then am creating a new class method called `submit_test` that is essentially a copy of the parent `submit` call with some modifications.

The parent's `submit` function looks like so:
```python
def submit(self, fn, /, *args, **kwargs):
    with self._shutdown_lock:
        if self._broken:
            raise BrokenProcessPool(self._broken)
        if self._shutdown_thread:
            raise RuntimeError('cannot schedule new futures after shutdown')
        if _global_shutdown:
            raise RuntimeError('cannot schedule new futures after '
                               'interpreter shutdown')

        f = _base.Future()
        w = _WorkItem(f, fn, args, kwargs)

        self._pending_work_items[self._queue_count] = w
        self._work_ids.put(self._queue_count)
        self._queue_count += 1
        # Wake up queue management thread
        self._executor_manager_thread_wakeup.wakeup()

        self._adjust_process_count()
        self._start_executor_manager_thread()
        return f
```

The question is why for each variable that uses the `self` keyword that is inherited from the parent class is giving the following error:
```
Cannot access member "X" for type "pool_test"
Member "X" is unknown (reportGeneralTypeIssues)
```

This error does not appear when using the `mypy` module, but still occurs when manually specifying a path to a cloned `typeshed` repository. This leads me to believe that `pyright` isn't able to see or use the `typeshed` information for the standard library, or that the `typeshed` data for `concurrent.futures` is not updated, but I'm not 100% sure.

**Comment by JelleZijlstra** [comment, marked as answer]
Mypy wouldn't give errors here because your function doesn't have type annotations. The issue is simply that you're using private attributes that aren't present in typeshed (https://github.com/python/typeshed/blob/master/stdlib/concurrent/futures/process.pyi). Typeshed's policy is to add private attributes when a user asks for them, so feel free to submit a PR to typeshed to add what you need.

**Reply by erictraut** [reply]
Yes, Jelle's explanation is correct.

I'm surprised to hear that typeshed maintainers would entertain the inclusion of private members in stdlib class definitions. Private members are internal implementation details that are not part of the public interface contract. They are not documented, and they could easily change from one version of Python to the next. Code that relies on private members will likely be fragile and hard to maintain.

**Reply by JelleZijlstra** [reply]
The motivation for this decision is in https://github.com/python/typeshed/issues/1902.
